viewer/Viewer.cpp

```diff
@@ -784,62 +989,141 @@ namespace Enki
+		int buttonCode = 0;
+		if(event->buttons() & Qt::LeftButton) buttonCode = 1;
+		else if(event->buttons() & Qt::RightButton) buttonCode = 2;
+		else if(event->buttons() & Qt::MiddleButton) buttonCode = 3;
+		else if(event->buttons() & Qt::MidButton) buttonCode = 3;
+
+		// if selected object is a robot call the clicked interaction function
+		Robot* robot = dynamic_cast<Robot*>(pointedObject);
+		if(robot) robot->clickedInteraction(true,buttonCode,pointedPoint.x(),pointedPoint.y(),pointedPoint.z());
 	}
 	
 	void ViewerWidget::mouseReleaseEvent(QMouseEvent * event)
 	{
-		/*if (event->button() == Qt::RightButton)
-			mouseGrabbed = false;*/
+		// enable physics calculation for selected object
+		world->SkipPhysicsObjectsList.erase(selectedObject);
+
+		// code button identification for robot
+		int buttonCode = 0;
+		if(event->buttons() & Qt::LeftButton) buttonCode = 1;
+		else if(event->buttons() & Qt::RightButton) buttonCode = 2;
+		else if(event->buttons() & Qt::MiddleButton) buttonCode = 3;
+		else if(event->buttons() & Qt::MidButton) buttonCode = 3;
+
+		// if selected object is a robot call the clicked interaction function
+		Robot* robot = dynamic_cast<Robot*>(pointedObject);
+		if(robot) robot->clickedInteraction(false,buttonCode,pointedPoint.x(),pointedPoint.y(),pointedPoint.z());
 	}
 	
 	void ViewerWidget::mouseMoveEvent(QMouseEvent *event)
 	{
-		if (event->modifiers() & Qt::ControlModifier)
+		// rotate selected object
+		if ((event->buttons() & Qt::LeftButton) && (event->buttons() & Qt::RightButton))
 		{
-			QPoint diff = event->pos() - mouseGrabPos;
-			if (event->buttons() & Qt::LeftButton)
+			if (isMovableByPicking(selectedObject))
 			{
-				if (event->modifiers() & Qt::ShiftModifier)
-				{
-					camera.pos.rx() -= 0.5 * cos(-camera.yaw) * (double)diff.y() + 0.5 * sin(-camera.yaw) * (double)diff.x();
-					camera.pos.ry() -= 0.5 * sin(-camera.yaw) * -(double)diff.y() + 0.5 * cos(-camera.yaw) * (double)diff.x();
-				}
-				else
-				{
-					camera.yaw -= 0.01 * (double)diff.x();
-					camera.pitch = clamp(camera.pitch - 0.01 * (double)diff.y(), -M_PI / 2, M_PI / 2);
-				}
+				QPoint diff = event->pos() - mouseGrabPos;
+				selectedObject->angle -= 0.01 * (double)diff.x();
+				mouseGrabPos = event->pos();
 			}
-			else if (event->buttons() & Qt::RightButton)
+		}
+
+		// move selected object if it's movable by picking
+		else if (event->buttons() & Qt::LeftButton)
+		{
+			if (isMovableByPicking(selectedObject) && (event->pos() - mouseGrabPos).manhattanLength() > 10 )
 			{
-				if (event->modifiers() & Qt::ShiftModifier)
+				if(!trackballView)
 				{
-					camera.altitude += -(double)diff.y();
+					world->SkipPhysicsObjectsList.insert(selectedObject);
+
+					selectedObject->pos = Point(pointedPoint.x(),pointedPoint.y());
+					selectedObject->speed = Vector(0,0);
+					selectedObject->angSpeed = 0;
 				}
 				else
 				{
-					// TODO: zoom
+					if(!messageList.empty() && messageList.back().first == controlError1)
```

Comment:
Enki's coding convention is to put a space between `if` and the opening parenthesis: `if (`.